packages/lambda-powertools-middleware-correlation-ids/index.js

```diff
@@ -1,298 +1,25 @@
-
-        if (!correlationIds[DEBUG_LOG_ENABLED]) {
-          correlationIds[DEBUG_LOG_ENABLED] = Math.random() < sampleDebugLogRate ? 'true' : 'false'
-        }
-
-        correlationIds[CALL_CHAIN_LENGTH] = (correlationIds[CALL_CHAIN_LENGTH] || 0) + 1
-
-        const correlationIdsInstance = new CorrelationIds(correlationIds)
-
-        Object.defineProperties(event, {
-          correlationIds: {
-            value: correlationIdsInstance,
-            enumerable: false
-          },
-          logger: {
-            value: new Log({ correlationIds: correlationIdsInstance }),
-            enumerable: false
-          }
-        })
-
-        return event
-      } catch (e) {
-        return initCorrelationIds(context, sampleDebugLogRate)
-      }
-    })
-
-  context.parsedKinesisEvents = events
-
-  // although we're going to have per-record correlation IDs, the default one for the function
-  // should still have the awsRequestId at least
-  CorrelationIds.replaceAllWith({
-    'x-correlation-id': awsRequestId,
-    awsRequestId,
-    [DEBUG_LOG_ENABLED]: Math.random() < sampleDebugLogRate ? 'true' : 'false'
-  })
-}
-
-function captureContextField ({ __context__ }, { awsRequestId }, sampleDebugLogRate) {
-  const correlationIds = __context__ || {}
-  correlationIds.awsRequestId = awsRequestId
-  if (!correlationIds[X_CORRELATION_ID]) {
-    correlationIds[X_CORRELATION_ID] = awsRequestId
-  }
-
-  if (!correlationIds[DEBUG_LOG_ENABLED]) {
-    correlationIds[DEBUG_LOG_ENABLED] = Math.random() < sampleDebugLogRate ? 'true' : 'false'
-  }
-
-  correlationIds[CALL_CHAIN_LENGTH] = (correlationIds[CALL_CHAIN_LENGTH] || 0) + 1
-
-  CorrelationIds.replaceAllWith(correlationIds)
-}
-
-function initCorrelationIds ({ awsRequestId }, sampleDebugLogRate) {
-  const correlationIds = { awsRequestId }
-  correlationIds[X_CORRELATION_ID] = awsRequestId
-  correlationIds[DEBUG_LOG_ENABLED] = Math.random() < sampleDebugLogRate ? 'true' : 'false'
-  correlationIds[CALL_CHAIN_LENGTH] = 1
-
-  CorrelationIds.replaceAllWith(correlationIds)
-}
-
-function isApiGatewayEvent (event) {
-  return event.hasOwnProperty('httpMethod')
-}
-
-function isSnsEvent (event) {
-  if (!event.hasOwnProperty('Records')) {
-    return false
-  }
-
-  if (!Array.isArray(event.Records)) {
-    return false
-  }
-
-  return event.Records[0].EventSource === 'aws:sns'
-}
-
-function isSqsEvent (event) {
-  if (!event.hasOwnProperty('Records')) {
-    return false
-  }
-
-  if (!Array.isArray(event.Records)) {
-    return false
-  }
-
-  return event.Records[0].eventSource === 'aws:sqs'
-}
-
-function isKinesisEvent (event) {
-  if (!event.hasOwnProperty('Records')) {
-    return false
-  }
-
-  if (!Array.isArray(event.Records)) {
-    return false
-  }
-
-  return event.Records[0].eventSource === 'aws:kinesis'
-}
-
-function hasContextField (event) {
-  return event.hasOwnProperty('__context__')
-}
+const generic = require('./event_sources/generic')
+const eventSources = [
```

Comment:
Really like splitting the event sources out too.

Could this be a map from event source name to a function to lazy load the needed event source?

Reply:
not sure I follow, wanna pseudo-code what you're thinking?

are you saying the user needs to say what event source to enable?

Reply:
@ChrisWestcottUK I like the idea of lazy-loading, but with this implementation we need to call `isMatch` to determine whether to run it. I think the only way around that would be to have one file which determines which event source it is, then lazy load after that.

Something along the lines of...

```js
const eventSourceDetector = require('./event-source-detector');
const getEventSource = event => {
  const source = eventSourceDetector(event);
  if (!source) return require('./event-sources/generic');
  return require(`./event-sources/${source}`);
});
// ...
const eventSource = getEventSource(event);
eventSource.captureCorrelationIds(event, context, sampleDebugLogRate);
```

`eventSourceDetector` would basically combine the `isMatch` functions from all the sources in one function, which isn't quite so nice to maintain but would have reasonable cold-start time improvements. `require` caching is very efficient, so we can rely on that for the lazy-loading.

Reply:
ah, I see what you meant now. Let's not prematurely optimize for cold start time until we know how much we stand to save. I'd wager that we won't make any measurable savings from lazy loading these few modules since they're all relative paths.

Reply:
Yeah all good, let's create an issue to investigate it as a potential enhancement